python/ray/data/_internal/logging.py

```diff
@@ -0,0 +1,64 @@
+import logging
+import os
+from typing import Optional
+
+import ray
+from ray._private.ray_constants import LOGGER_FORMAT
+
+_default_log_handler: logging.Handler = None
+
+DEFAULT_DATASET_LOG_FILENAME = "ray-data.log"
+
+
+class LogHandler(logging.Handler):
```

Comment:
For future maintainers, I think it would be really valuable to have some documentation so they know

1. What this log handler is doing (why is there a special handler in Ray Data? There are no other Ray libraries with special logging handlers, etc).
2. How and in what cases it is meant to be used (or if they should never touch it)
3. A note about how it relates to the main Ray logging system could also be really useful here

Reply:
Will add documentation.

> I'm also wondering why this cannot be done via configuration.

The Ray session directory isn't available until Ray initializes. So, we need to configure the logging path at runtime. If you configure the logging path directly in `configure_logging`, then we can't call the function until Ray has initialized. This makes things more complicated.

> What this log handler is doing  

Because the session directory isn't available when the module is loaded, this handler configures the logging path when a message is emitted rather than when the handler is created. 

> why is there a special handler in Ray Data? There are no other Ray libraries with special logging handlers, etc

Both Serve and Data write logs to the Ray session directory. Serve only writes logs in a couple of places (proxy and replica actors) where you can assume Ray is initialized. In contrast, Ray Data writes logs throughout the codebase, and we can't always assume Ray is initialized. 

> How and in what cases it is meant to be used (or if they should never touch it)

 Should never touch it. 

> A note about how it relates to the main Ray logging system could also be really useful here

The main Ray logging system configures writing to console. This handler is for writing to files.